### PR TITLE
Support conditional setting of serialization_scope in controllers

### PR DESCRIPTION
**Pull request by yogeshjain999**
#### Purpose
I want to set `serialization_scope` as different object rather than `current_user`, for only specific actions in given controller.

#### Changes
Adds support for setting `serialization_scope` conditionally in controllers, same as calling `before_action` with options.

### CODE REVIEW

**Comment by bf4** [comment]
I like it.  I need to look at the test changes a bit.

File: test/action_controller/serialization_scope_name_test.rb [hunk 1]
```diff
@@ -71,10 +77,12 @@ class DefaultScopeTest < ActionController::TestCase
     tests PostTestController
 
     def test_default_serialization_scope
+      get :render_post
```
Comment (bf4) [inline test/action_controller/serialization_scope_name_test.rb]: this is supposed to be missing :)

File: test/action_controller/serialization_scope_name_test.rb [hunk 2]
```diff
@@ -187,14 +233,6 @@ def new_post
...
     tests PostViewContextTestController
 
-    def test_nil_serialization_scope
-      assert_nil @controller._serialization_scope
-    end
-
-    def test_nil_serialization_scope_object
-      assert_nil @controller.serialization_scope
-    end
-
```
Comment (bf4) [inline test/action_controller/serialization_scope_name_test.rb]: Unrelated change, no?

File: lib/action_controller/serialization.rb [hunk 3]
```diff
@@ -8,8 +8,10 @@ module Serialization
     include ActionController::Renderers
 
     module ClassMethods
-      def serialization_scope(scope)
-        self._serialization_scope = scope
+      def serialization_scope(scope, opts = {})
+        before_action(opts.slice(:only, :except, :if, :unless)) do
```
Comment (bf4) [inline lib/action_controller/serialization.rb]: Since the method is only called at `render`, if not passed in as an option, I don't think I like having it eagerly evaluated in code that is pretty easy to handle in userland.  Would you be okay with removing the code changes in favor documenting using a `before_action` as an alternative to passing an option to `render`?

Comment (bf4) [inline lib/action_controller/serialization.rb]: I also would except there to be bugs with `current_user` being called in a `before_action` with all the side-effects authentication-related libs can have

File: test/action_controller/serialization_scope_name_test.rb [hunk 4]
```diff
@@ -34,6 +34,12 @@ def json_key
   end
   class PostTestController < ActionController::Base
     attr_accessor :current_user
+
+    def render_post
+      self.current_user = User.new(id: 3, name: 'Pete')
+      render json: new_post, serializer: serializer, adapter: :json
+    end
```
Comment (bf4) [inline test/action_controller/serialization_scope_name_test.rb]: this condition is identical to the one below it :)

**Comment by bf4** [comment]
Closing as stale. No response in a month.

File: lib/action_controller/serialization.rb (hunk 3 above)
Comment (yogeshjain999) [inline lib/action_controller/serialization.rb]: I agree to your suggestion @bf4 about moving this to userland, created #1824 with added documentation.
